File: java/RadixSort.java
// noncomparative sorting algorithm that sorts numbers digit by digit starting from the least
// significant digit to the most significant digit 

public class RadixSort {
  public static void radixSort(int[] arr) {
    int max = getMax(arr, arr.length);

    for(int exp = 1; max / exp > 0; exp *= 10) {
      countSort(arr, n, exp);
    }
  }
  static int getMax(int arr[], int n) {
    int max = arr[0];
    for(int i = 1; i < n; i++)
      if(arr[i] > max)
        max = arr[i];
    return max;
  }
  // function to do coutning sort of arr[] according to digit represetned by exponent
  static void countSort(int arr[], int n, int exp) {
    int output[] = new int[n];
    int i;
    int count[] = new int[10];
    Arrays.fill(count, 0);

    for(i = 0; i < n; i++)
      count[(arr[i] / exp) % 10]++;

    for(i = 1; i < 10; i++)
      count[i] += count[i - 1];

    // build output array
    for(i = n - 1; i >= 0; i--) {
      output[count[(arr[i] / exp) % 10] - 1] = arr[i];
      count[(arr[i] / exp) % 10]--;
    }
    for(i = 0; i < n; i++)
      arr[i] = output[i];
  }
}

File: c/radixSort.c
#include <stdio.h>

void radixSort(int arr[], int n) {
  // find max number to knwo number of digits
  int max = arr[0];
  for(int i = 1; i < n; i++)
    if(arr[i] > max)
      max  = arr[i];

  // do counting sort for every digit; instead of passing digit number, exp is passed 
  // exp is 10 ^ i where i is current digit number
  for(int exp = 1; max / exp > 0; exp *= 10) {
    int output[n];
    int count[10] = {0};

    // store count of occurrences in count[]
    for(int i = 0; i < n; i++)
      count[(arr[i] / exp) % 10]++;

    // change count[i] so it contains actual position of this digit in output
    for(int i = 1; i < 10; i++)
      count[i] += count[i - 1];

    // build output array
    for(int i = n - 1; i >= 0; i--) {
      output[count[(arr[i] / exp) % 10] - 1] = arr[i];
      count[(arr[i] / exp) % 10]--;
    }
    // copy output array to arr[]
    for(int i = 0; i < n; i++)
      arr[i] = output[i];
  }
}

// function to print array
void printArray(int arr[], int n) {
  for(int i = 0; i < n; i++)
    printf("%d ", arr[i]);
}

int main() {
  int arr[] = {170, 45, 75, 90, 802, 24, 2, 66};
  int n = sizeof(arr) / sizeof(arr[0]);

  radixSort(arr, n);
  printArray(arr, n);
  return 0;
}
